### PR TITLE
Add image driver interface, configuration and plugin callback

### PR DESCRIPTION
**Pull request by cclerget**
## Description of the Pull Request (PR):

- Add an image driver interface to handle image mount setup in place of loop devices and optionally overlay
- Added to `singularity.conf`:
  - new `image driver` directive
  - new `driver` value for `enable overlay`
- Plugin callback to register image driver
- Few fixes and a bit of refactoring for fuse mount code
- Add a plugin example to enable overlay mount with user namespaces

#### Before submitting a PR, make sure you have done the following:

- Read the [Guidelines for Contributing](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md), and this PR conforms to the stated requirements.
- Added changes to the [CHANGELOG](https://github.com/sylabs/singularity/blob/master/CHANGELOG.md) if necessary according to the [Contribution Guidelines](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md)
- Added tests to validate this PR and tested this PR locally with a `make testall`
- Based this PR against the appropriate branch according to the [Contribution Guidelines](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md)
- Added myself as a contributor to the [Contributors File](https://github.com/sylabs/singularity/blob/master/CONTRIBUTORS.md)

Attn: @singularity-maintainers

### CODE REVIEW

**Comment by dtrudg** [comment]
Is the Travis failure here now something we aren't covering with Circle due to limiting to short integration tests.... and we aren't hitting through e2e?

```
  conveyorPacker_oci_test.go:269: failed to Get from docker://alpine: Error initializing source oci:/tmp/image_cache-620861570/cache/oci:ab00606a42621fb68f2ed6ad3c88be54397f981a7b70a79db3d1172b11c4367d: Error creating an updated image manifest: Get https://registry-1.docker.io/v2/library/alpine/blobs/sha256:e7d92cdc71feacf90708cb59182d0df1b911f8ae022d29e8e95d75ca6a99776a: EOF

FAIL

coverage: 23.5% of statements
```

I'm a little concerned about missing something in Circle now.

**Comment by cclerget** [comment]
> Is the Travis failure here now something we aren't covering with Circle due to limiting to short integration tests.... and we aren't hitting through e2e?

Looks like a Travis network error (or docker hub ?), this error came from unit tests, also same error during integration tests with few tests pulling from docker.

File: pkg/util/singularityconf/config.go [hunk 1]
```diff
@@ -323,4 +325,11 @@ mksquashfs procs = {{ .MksquashfsProcs }}
 # Allow to share same images associated with loop devices to minimize loop
 # usage and optimize kernel cache (useful for MPI)
 shared loop devices = {{ if eq .SharedLoopDevices true }}yes{{ else }}no{{ end }}
+
+# IMAGE DRIVER: [STRING]
+# DEFAULT: Undefined
+# Image driver allows to specify a driver installed by a plugin responsible
```
Comment (dtrudg) [inline pkg/util/singularityconf/config.go]: How about...

This option specifies the name of an image driver provided by a plugin that will be used to handle image mounts. If the 'enable overlay' option is set to 'driver'  the driver name specified here will also be used to handle overlay mounts. If the driver name specified has not been registered via a plugin installation the run-time will abort.

**Review by dtrudg — COMMENTED** [review]
I have one suggested change on the config file explanatory text.

I'm going to check the overlay driver on Ubuntu this morning, and expect to approve then.

**Comment by dtrudg** [comment]
On Ubuntu 19.10 I have installed the driver and then hit this:

```
dave@lappy:~
08:00 AM $ singularity shell -u --overlay overlay.img ubuntu_latest.sif 
INFO:    Convert SIF file to sandbox...
INFO:    Cleaning up image...
FATAL:   container creation failed: mount /proc/self/fd/7->/usr/local/var/singularity/mnt/session/overlay-images/0 error: while mounting image /proc/self/fd/7: failed to find loop device: could not attach image file to loop device: permission denied
dave@lappy:~
08:00 AM $ cat /etc/lsb-release 
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=19.10
DISTRIB_CODENAME=eoan
DISTRIB_DESCRIPTION="Ubuntu 19.10"
```

**Comment by dtrudg** [comment]
Does work with directory overlays, of course loop mount isn't available without priv - agree with discussion in chat that a nicer error message would be good.

```
08:04 AM $ singularity shell -u --overlay overlay-dir ubuntu_latest.sif 
INFO:    Convert SIF file to sandbox...
Singularity> df -h
Filesystem                       Size  Used Avail Use% Mounted on
overlay                          227G   35G  181G  16% /
udev                             5.8G     0  5.8G   0% /dev
tmpfs                            5.8G   46M  5.8G   1% /dev/shm
/dev/mapper/vgubuntu--mate-root  227G   35G  181G  16% /tmp
tmpfs                            5.8G     0  5.8G   0% /sys/fs/cgroup
tmpfs                             16M   12K   16M   1% /etc/group
Singularity> 
```

**Review by dtrudg — APPROVED** [review]

**Pull request merged**